### PR TITLE
Add logic to block users access to confirmation page

### PR DESCRIPTION
**Pull request by sebastien-comeau**
### Description

Add logic to block users access to confirmation page if application have not been submitted. The inverse also applied, if the users have sumitted their applications then only confirmation page is accessible.

### Checklist

- [x] I have tested the changes locally
- [x] I have updated the documentation if necessary
- [ ] I have added/updated tests that prove my fix is effective or that my feature works
- [x] I have checked that my code follows the project's coding style by running `npm run format:check`
- [x] I have checked that my code contains no linting errors by running `npm run lint`
- [x] I have checked that my code contains no type errors by running `npm run typecheck`
- [x] I have checked that all unit tests pass by running `npm run test:unit -- run`
- [x] I have checked that all e2e tests pass by running `npm run test:e2e`

### Test Instructions

1. Try to access `/apply/:id/confirmation` page while completing the application. You should be redirected to first page of the flow `/apply/:id/type-of-application`.
2. Complete and submit the application and once you're on the `/apply/:id/confirmation` page, try to access any flow pages. You should be redirected to `/apply/:id/confirmation` page.

### CODE REVIEW

File: frontend/app/routes-flow/apply-flow.ts [hunk 1]
```diff
@@ -162,7 +200,9 @@ async function start({ id, request }: StartArgs) {
 
   const sessionName = getSessionName(parsedId);
   session.set(sessionName, initialState);
-  session.set('timeUpdated', Date.now());
+
+  const timeUpdatedSessionName = getTimeUpdatedSessionName(id);
+  session.set(timeUpdatedSessionName, new Date().toISOString());
```
Comment (gregory-j-baker) [inline frontend/app/routes-flow/apply-flow.ts:205]:
```suggestion
  session.set(timeUpdatedSessionName, new Date());
```

Reason:
The date will be `JSON.stringify()'d` when storing to session, which will automatically convert it to an ISO string.

File: frontend/app/routes/$lang+/_public+/apply+/$id+/review-information.tsx [hunk 2]
```diff
@@ -6,6 +6,7 @@ import { useFetcher, useLoaderData } from '@remix-run/react';
 
 import { faSpinner, faX } from '@fortawesome/free-solid-svg-icons';
 import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { randomBytes } from 'crypto';
```
Comment (gregory-j-baker) [inline frontend/app/routes/$lang+/_public+/apply+/$id+/review-information.tsx:9]:
```suggestion
import { randomBytes } from 'node:crypto';
```

Reference:
- <https://nodejs.org/docs/latest-v20.x/api/crypto.html#cryptorandombytessize-callback> (switch it to ESM syntax and look at example)

**Review by gregory-j-baker — CHANGES_REQUESTED** [review]

File: frontend/app/routes-flow/apply-flow.ts (hunk 1 above)
Comment (sebastien-comeau) [reply frontend/app/routes-flow/apply-flow.ts:205]: But it means that I need to parse it as Date when reading it.

File: frontend/app/routes-flow/apply-flow.ts [hunk 3]
```diff
@@ -37,49 +40,80 @@ interface ApplyState {
...
 
   const state = session.get(sessionName) as ApplyState;
 
-  if (Date.now() - session.get('timeUpdated') > 15 * 60 * 1000) {
+  // Checks if the elapsed time since the last update exceeds 15 minutes,
+  // and performs necessary actions if it does.
+  const timeUpdatedSessionName = getTimeUpdatedSessionName(id);
+  const timeUpdatedSession = session.get(timeUpdatedSessionName) as string;
+  const timeUpdated = new Date(timeUpdatedSession);
+  const now = new Date(new Date().toISOString());
```
[the remark this answers is not in the document]
Comment (sebastien-comeau) [reply frontend/app/routes-flow/apply-flow.ts]: Fixed

File: frontend/app/routes-flow/apply-flow.ts [hunk 4]
```diff
@@ -37,49 +40,80 @@ interface ApplyState {
...
+
   const sessionService = await getSessionService();
   const session = await sessionService.getSession(request);
-
-  const sessionName = getSessionName(id.data);
+  const sessionName = getSessionName(id);
 
   if (!session.has(sessionName)) {
-    throw redirectWithLocale(request, fallbackRedirectUrl, 302);
+    throw redirectWithLocale(request, applyRouteUrl, 302);
```
[the remark this answers is not in the document]
Comment (sebastien-comeau) [reply frontend/app/routes-flow/apply-flow.ts]: Fixed

File: frontend/app/routes-flow/apply-flow.ts [hunk 5]
```diff
@@ -37,49 +40,80 @@ interface ApplyState {
...
-  const id = idSchema.safeParse(params.id);
+async function loadState({ params, request }: LoadStateArgs) {
+  const applyRouteUrl = '/apply';
+  const { pathname } = new URL(request.url);
+  const parsedId = idSchema.safeParse(params.id);
 
-  if (!id.success) {
-    throw redirectWithLocale(request, fallbackRedirectUrl, 302);
+  if (!parsedId.success) {
+    throw redirectWithLocale(request, applyRouteUrl, 302);
```
[the remark this answers is not in the document]
Comment (sebastien-comeau) [reply frontend/app/routes-flow/apply-flow.ts]: Fixed

File: frontend/app/routes-flow/apply-flow.ts [hunk 6]
```diff
@@ -37,49 +40,80 @@ interface ApplyState {
...
       },
     });
   }
 
-  return { id: id.data, state };
+  // Redirect to the confirmation page if the application has been submitted and
+  // the current route is not the confirmation page.
+  const confirmationRouteUrl = `/apply/${id}/confirmation`;
+  if (state.submissionInfo && !pathname.endsWith(confirmationRouteUrl)) {
+    throw redirectWithLocale(request, confirmationRouteUrl, 302);
```
[the remark this answers is not in the document]
Comment (sebastien-comeau) [reply frontend/app/routes-flow/apply-flow.ts]: Fixed

File: frontend/app/routes-flow/apply-flow.ts [hunk 7]
```diff
@@ -37,49 +40,80 @@ interface ApplyState {
...
+  }
+
+  // Redirect to the first flow page if the application has not been submitted and
+  // the current route is the confirmation page.
+  if (!state.submissionInfo && pathname.endsWith(confirmationRouteUrl)) {
+    throw redirectWithLocale(request, `/apply/${id}/type-of-application`, 302);
```
[the remark this answers is not in the document]
Comment (sebastien-comeau) [reply frontend/app/routes-flow/apply-flow.ts]: Fixed

File: frontend/app/routes-flow/apply-flow.ts [hunk 8]
```diff
@@ -107,7 +141,9 @@ async function saveState({ params, request, state, remove = undefined }: SaveSta
 
   const sessionName = getSessionName(id);
   session.set(sessionName, newState);
-  session.set('timeUpdated', Date.now());
+
+  const timeUpdatedSessionName = getTimeUpdatedSessionName(id);
+  session.set(timeUpdatedSessionName, new Date().toISOString());
```
[the remark this answers is not in the document]
Comment (sebastien-comeau) [reply frontend/app/routes-flow/apply-flow.ts:146]: Explicitly assigning it as a string, despite the eventual conversion of Date objects to strings, could assist developers. There's a possibility they might attempt to access the session and invoke date functions on it, leading to errors, whereas ensuring it's a string would return the expected result.

[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply frontend/app/routes-flow/apply-flow.ts:146]: Here's an improved version of your pull request comment:

**Great catch!** By explicitly converting the data to a string before calling `session.set()`, we improve code clarity for future developers. This makes it clear that the data is stored as a string in the session and needs to be parsed back into its original format when retrieved with `session.get()`.

**Here's why this matters:**

* **Consistency:** Ensures consistent data handling throughout the codebase.
* **Maintainability:** Makes the code easier to understand and maintain for future developers.
* **Reduces Errors:** Helps prevent potential errors caused by accidentally treating the data as a different type.

**This is a good practice to follow!**

File: frontend/app/routes-flow/apply-flow.ts (hunk 1 above)
Comment (gregory-j-baker) [reply frontend/app/routes-flow/apply-flow.ts:205]: See <https://github.com/DTS-STN/canadian-dental-care-plan/pull/803#discussion_r1539016895>

**Pull request merged**